plugins/framework/tests/test_model_patcher.py

```diff
@@ -0,0 +1,136 @@
+    # 1. Register rule and specify a trigger on target module for the rule to be applied
+    # 2. Patch model
+    # 3. check target module's forward function and dummy patch produces similar outputs
+    dummy_forward_to_patch = lambda self, X: X * 2 # pylint: disable=unnecessary-lambda-assignment
+    ModelPatcher.rules.pop(DUMMY_RULE_ID, None)
+    rule = ModelPatcherRule(
+        rule_id=DUMMY_RULE_ID,
+        trigger=ModelPatcherTrigger(check=DummyModule),
+        forward=dummy_forward_to_patch,
+    )
+    ModelPatcher.register(rule)
+    assert DUMMY_RULE_ID in ModelPatcher.rules.keys(), "Rule Registration Failed" # pylint: disable=consider-iterating-dictionary
+    ModelPatcher.patch(model)
+    assert torch.allclose(
+        model(model_inputs), model_inputs * 2
+    ), "Failed to patch forward function"
+
+
+# Test patching of model attribute
+def test_patching_downstream_module(model): # pylint: disable=redefined-outer-name
+    "Test patching an imported module indirectly managed by other modules using import_and_reload"
+    # 1. Register rule targeting downstream module and specify target to reload with patch applied
+    # 2. Patch model
+    # 3. check patched module now exist in model
+    ModelPatcher.rules.pop(DUMMY_RULE_ID, None)
+
+    # Reload because we only want to patch CrossEntropyLoss for this target module
+    ModelPatcher.register(
+        ModelPatcherRule(
+            rule_id=DUMMY_RULE_ID,
+            import_and_maybe_reload=(
+                "torch.nn.CrossEntropyLoss",
+                DummyCrossEntropyLoss,
+                "fms_acceleration.utils.test_utils",
+            ),
+        )
+    )
+    ModelPatcher.patch(model)
+    assert isinstance(
+        DummyModule(DUMMY_HIDDEN_DIM).loss_fn, DummyCrossEntropyLoss
+    ), "Failed to patch attribute with import and reload"
+
+
+# Test patching standalone functions
+def test_patching_standalone_function(model_inputs): # pylint: disable=redefined-outer-name
+    "Test patching of standalone file functions"
+    # 1. Take an arbitrary function
+    # 2. replace with a dummy function
+    # 3. check that the arbitrary function and dummy functions produces similar outputs
+    dummy_function_to_patch = lambda X: X # pylint: disable=unnecessary-lambda-assignment
+    patch_target_module(
+        "fms_acceleration.utils.test_utils.read_configuration",
+        dummy_function_to_patch,
+    )
+    # First Party
+    from fms_acceleration.utils.test_utils import read_configuration # pylint: disable=import-outside-toplevel
+
+    assert torch.allclose(
+        read_configuration(model_inputs), model_inputs
+    ), "Failed to patch standalone function"
+
+
+def test_forward_patching_with_forward_builder(model_inputs, model): # pylint: disable=redefined-outer-name
+    "Test model patcher replaces forward using a dummy forward building function"
+
+    def dummy_forward_builder(module, multiplier):
+        # can apply modifications to module here
```

Comment:
this one you should do it like a spy. see the spy mocks in fms-hf-training https://github.com/foundation-model-stack/fms-hf-tuning/blob/main/tests/acceleration/test_acceleration_framework.py#L74-L78